### ISSUE TITLE
Faster hashing

### ISSUE THREAD

**Issue by anntzer** [joblib/joblib#343]
joblib currently defaults to md5-hashing its input.  For the tasks at hand, a non-cryptographic hash can be significantly faster (see comparison table at http://cyan4973.github.io/xxHash/).
scikit-learn provides an implementation of murmurhash and there's an xxhash implementation available on PyPI.

**Comment by lesteve** [comment]
Do you have any benchmarks that would show it is worthwile for joblib hashes?

A problem is that we try to reduce joblib dependencies as much as possible, in particular dependencies that require compilation (outside of numpy) is a big no no.

**Comment by ogrisel** [comment]
I ran some benchmarks in the past and as I recall, md5sum can hash at a couple hundred MB/s while xxhash can typically hash at close to the memory bandwidth speed (11GB/s on my workstation), so much much faster.

However joblib aims to stay a pure-python project. We could maybe support alternative hashing algorithms if the user accepts to install optional dependencies by themselves. I think we need to change the way we store our digests to include the name of the hashing algorithm as a prefix, e.g.:

md5:deadbeefcafebabe
xxh64:deadbeefcafebabe

This way if joblib finds a cache entry keyed by digest with a hashing algorithm that is not installed, it can raise a user warning to let the user know about it (and proceed with re-computation as if the cached entry did not exist).

**Comment by GaelVaroquaux** [comment]
> I ran some benchmarks in the past and as I recall, md5sum can hash at a couple
> hundred MB/s while xxhash can typically hash at close to the memory bandwidth
> speed (11GB/s on my workstation), so much much faster.

Well, I think that the question really is: in a typical situation, how
much is the overall usage of joblib limited by the time the hashing
algorithm takes.

**Comment by ogrisel** [comment]
It's true that if the data is only on disk (cold cache), the md5sum speed matches approximately the speed of typical high end HDD (although not the latest NVMe SSD).

If you have a 10GB array and pass it to a function with a `Memory.cache` decorator, just checking that you have nothing to do will take close to 20s with md5 while it could be just 1s with xxh64.

**Comment by rth** [comment]
Also, it looks like `joblib.hash` is much slower for DataFrames and sparse arrays than [`dask.base.tokenize`](https://github.com/dask/dask/pull/742)  which is also using md5 as far as I understand.  (I'm wondering if dask tokenization is valid for sparse arrays though).

In the benchmarks below it is 30x slower for hashing pandas DataFrames and 100x times slower for sparse numpy arrays (though I'm not sure if running benchmarks with multiple iterations of `timeit` could have an effect here),

```py
import pandas as pd
import numpy as np
import scipy.sparse

from dask.base import tokenize
from sklearn.externals import joblib

df = pd.read_csv('http://www.gagolewski.com/resources/data/titanic3.csv', comment='#')

rng = np.random.RandomState(42)
X = rng.rand(1000, 100)
X_csr = scipy.sparse.rand(1000, 10000, random_state=rng)

for data in [df, X, X_csr]:

    print('# {}, shape={}'.format(type(data).__name__, data.shape))
    print('   joblib.hash          ', end='')
    %timeit joblib.hash(data)
    print('   dask.base.tokenize   ', end='')
    %timeit tokenize(data)
```
which produces,
```
# DataFrame, shape=(1309, 14)
   joblib.hash          33.5 ms ± 3.71 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
   dask.base.tokenize   3.67 ms ± 95.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# ndarray, shape=(1000, 100)
   joblib.hash          1.63 ms ± 12.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
   dask.base.tokenize   1.45 ms ± 254 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
# csr_matrix, shape=(1000, 10000)
   joblib.hash          3.58 ms ± 781 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
   dask.base.tokenize   26.6 µs ± 2.43 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

>> A problem is that we try to reduce joblib dependencies as much as possible, in particular dependencies that require compilation (outside of numpy) is a big no no.
>
> We could maybe support alternative hashing algorithms if the user accepts to install optional dependencies by themselves

Accepting custom hashing function would quite useful. In particular in the context of scikit-learn, mumurhash is already there...

**Comment by rth** [comment]
>  (I'm wondering if dask tokenization is valid for sparse arrays though)

It's not (https://github.com/dask/dask/issues/3054) so please ignore the results regarding sparse arrays, they are still relevant for pandas DataFrames though.

**Comment by lesteve** [comment]
Of course any hint on why joblib is a lot slower would be more than welcome. For example, is it slower on numpy arrays as well? Is it slower only on DataFrames?

**Comment by lesteve** [comment]
From what I understand from https://github.com/dask/dask/pull/742/files#diff-e5d233d0b11c3223493928fdb5fb35ceR155, it seems like dask is hashing only some of the attributes whereas joblib will walk the object with pickle and hash all of its attributes. Not sure whether that explains the difference.

**Comment by rth** [comment]
> Of course any hint on why joblib is a lot slower would be more than welcome. For example, is it slower on numpy arrays as well? Is it slower only on DataFrames?

Yes, only DataFrames as far as I can tell. Dense and sparse arrays have a very similar performance, which is reassuring.

> it seems like dask is hashing only some of the attributes whereas joblib will walk the object with pickle and hash all of its attributes.

Yes, it has a custom hashing (or in their vocabulary "tokenization" as it has somewhat different goals https://github.com/dask/dask/issues/3054#issuecomment-356638907) implementation for each of the supported objects. The joblib approach is indeed much more general.

**Comment by rth** [comment]
Another thing we could do it to fallback to `pandas.util.hash_pandas_object` for pandas object. It's much faster. With the above dataset I get,
```py
%timeit joblib.hash(df)
# 12.1 ms ± 376 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit pd.util.hash_pandas_object(df)
# 2.38 ms ± 62.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by judahrand** [comment]
For anyone who is still interested in this issue - the two PRs I've submitted dramatically improve the hashing speed for large DataFrame arguements for me.
